bluesky/callbacks/tiled_writer.py

```diff
@@ -0,0 +1,169 @@
+        parent_node = descriptor_node["external"]
+
+        num_rows = (
+            doc["indices"]["stop"] - doc["indices"]["start"]
+        )  # Number of rows added by new StreamDatum
+
+        # Get the Stream Resource node if it already exists or register if from a cached SR document
+        SR_uid = doc["stream_resource"]
+        if SR_uid in self._SR_nodes.keys():
+            SR_node = self._SR_nodes[SR_uid]
+        elif SR_uid in self._SR_cache.keys():
+            # Register a new (empty) Stream Resource
+            SR_doc = self._SR_cache.pop(SR_uid)
+
+            # POST /api/v1/register/{path}
+            file_path = (
+                "/"
+                + SR_doc["root"].strip("/")
+                + "/"
+                + SR_doc["resource_path"].strip("/")
+            )
+            data_path = SR_doc["resource_kwargs"]["path"].strip("/")
+            data_uri = "file://localhost" + file_path
+            assets = [
+                Asset(data_uri=data_uri, is_directory=False, parameter="data_uri")
+            ]
+            data_key = SR_doc["data_key"]
+            data_desc = dict(descriptor_node.metadata)["data_keys"][data_key]
+            if data_desc["dtype"] == "array":
+                data_shape = data_desc["shape"]
+            elif data_desc["dtype"] == "number":
+                data_shape = ()
+            data_type = np.dtype(
+                data_desc.get("dtype_str", "<f8")
+            )  # Find machine dtype; assume '<f8' by default
+
+            SR_node = parent_node.new(
+                structure_family=StructureFamily.array,
+                data_sources=[
+                    DataSource(
+                        assets=assets,
+                        mimetype=MIMETYPE_LOOKUP[SR_doc["spec"]],
+                        structure_family=StructureFamily.array,
+                        structure=ArrayStructure(
+                            data_type=BuiltinDtype.from_numpy_dtype(data_type),
+                            shape=[0, *data_shape],
+                            chunks=[[]] + [[d] for d in data_shape],
+                        ),
+                        parameters={"path": data_path.split("/")},
+                        management=Management.external,
+                    )
+                ],
+                metadata=SR_doc,
+                specs=[],
+            )
+            self._SR_nodes[SR_uid] = SR_node
+        else:
+            raise RuntimeError(
+                f"Stream Resource {SR_uid} is referenced before being declared."
+            )
+
+        # Append StreamDatum to an existing StreamResource (by overwriting it with changed shape)
+        url = SR_node.uri.replace("/metadata/", "/data_source/")
+        SR_node.refresh()
+        ds_dict = SR_node.data_sources()[0]
+        ds_dict["structure"]["shape"][0] += num_rows
+        ds_dict["structure"]["chunks"][0].append(num_rows)
+        is_chunked = SR_node.metadata["resource_kwargs"].get("chunked", True)
```

Comment:
Yes, `event_model` is where we keep schemas for the documents. The details of `resource_kwargs` (to be renamed `parameters`) will be different for each `spec` (to be renamed `mimetype`)---that is, each filetype. When we factor the HDF5-specific aspects of this into separate classes, I think those classes should define schemas and/or validation methods.

This idea has been [outstanding for awhile](https://github.com/bluesky/area-detector-handlers/issues/6) and it would be great to pick it up!

Regarding `chunked`: I think this _is_ the right place to put this. I'm wondering whether a boolean is too coarse. I think what we're specifying here is the chunk size along the time axis. Did you consider using `chunk_size: Optional[int]`, where you can give:

- `chunk_size=1` -- one HDF5 chunk per frame
- `chunk_size=N` -- N frames into each HDF5 chunk
- `chunk_size=None` -- all frames into one HDF5 chunk

Reply:
I actually thought about this too; I agree, boolean might be too restrictive. I wasn't sure if we also need to consider variable-size chunks, but I like the idea of `chunk_size` -- it's way less ambiguous. Perhaps we can assume for now that it's either 1, None, or fixed.

Reply:
Updated this to allow for variable-sized chunks as well.